AlreadyBored/shop-backend
# Task 9 solution

# Task 9 Backend for frontend
## [Task description](https://github.com/EPAM-JS-Competency-center/cloud-development-course-initial/blob/main/task9-bff-elastic-beanstalk/task.md)

## Evaluation criteria

Links:
* BFF service deployed on Elastic Beanstalk http://alreadybored-bff-api-default.eu-west-1.elasticbeanstalk.com/
* Cart service (without BFF)  https://79hwnkk528.execute-api.eu-west-1.amazonaws.com/default/api/profile/cart
* Cart service (from BFF) http://alreadybored-bff-api-default.eu-west-1.elasticbeanstalk.com/cart/api/profile/cart
* Product service (without BFF) https://icqa4iybt4.execute-api.eu-west-1.amazonaws.com/dev/products
* Product service (from BFF) http://alreadybored-bff-api-default.eu-west-1.elasticbeanstalk.com/products

* **3** - A working and correct **express** application should be in the **bff-service** folder. Reviewer can start this application locally with any valid configuration in the **.env** file and this application should works as described in the task 9.1 (last version of application uses Nest instead of Express but you can find working Express application by simply going to commit with hash **9efea40f74976dcaaced656187db46f4e4185687**) [V]
* **5** - The **bff-service** should be deployed with Elastic Beanstalk. The **bff-service** call should be redirected to the appropriate service : **product-service** or **CART**. The response from the **bff-service** should be the same as if **product-service** or **CART** services were called directly (look at links above) [V]
![bff-app](https://user-images.githubusercontent.com/34455330/140639512-2db55fb0-09ea-46b1-b5de-f8b57bf896e7.JPG)
![bff-env](https://user-images.githubusercontent.com/34455330/140639513-4e5c8670-75c5-48b1-a041-393273ef25b5.JPG)


 
## Additional (optional) tasks
---
* **+1** - Add a cache at the **bff-service** level for a request to the **getProductsList** function of the **product-service**. The cache should expire in 2 minutes [V]
How to test:
  * Get products list
  **GET** request to http://alreadybored-bff-api-default.eu-west-1.elasticbeanstalk.com/products
  * Create new product
  **POST** request to http://alreadybored-bff-api-default.eu-west-1.elasticbeanstalk.com/products with body
{
	"title": "PRODUCT_TITLE",
	"price": 22,
	"count": 33,
	"image": "LINK_TO_PRODUCT_IMAGE",
	"description": "PRODUCT_DESCRIPTION"
}
  * Get products list - result shouldn’t have new product
    **GET** request to http://alreadybored-bff-api-default.eu-west-1.elasticbeanstalk.com/products
  * Wait more than 2 minutes
  * Get products list - result should have new product
    **GET** request to http://alreadybored-bff-api-default.eu-west-1.elasticbeanstalk.com/products
* **+1** - Use **NestJS** to create **bff-service** instead of **express** (last version of application uses Nest instead of Express but you can find working Express application by simply going to commit with hash **9efea40f74976dcaaced656187db46f4e4185687**) [V]
---
Total score is 7/7
